state/etcd/etcd.go

```diff
@@ -0,0 +1,481 @@
+
+	var err error
+	if etag != nil {
+		etag, _ := strconv.ParseInt(*etag, 10, 64)
+		_, err = e.client.Txn(ctx).
+			If(clientv3.Compare(clientv3.ModRevision(key), "=", etag)).
+			Then(clientv3.OpDelete(key)).
+			Commit()
+	} else {
+		_, err = e.client.Delete(ctx, key)
+	}
+	if err != nil {
+		return fmt.Errorf("couldn't delete key %s: %s", key, err)
+	}
+	return nil
+}
+
+func (e *Etcd) doValidateEtag(key string, etag *string, concurrency string) error {
+	hasEtag := etag != nil && *etag != ""
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	if concurrency == state.FirstWrite && !hasEtag {
+		item, err := e.client.Get(ctx, key)
+		if err != nil {
+			return fmt.Errorf("couldn't get key %s: %s", key, err)
+		} else if item != nil {
+			return state.NewETagError(state.ETagMismatch, fmt.Errorf("item already exists and no etag was passed"))
+		} else {
+			return nil
+		}
+	} else if hasEtag {
+		item, err := e.client.Get(ctx, key)
+		if err != nil {
+			return fmt.Errorf("couldn't get key %s: %s", key, err)
+		}
+		if item == nil || item.Kvs == nil {
+			return state.NewETagError(state.ETagMismatch, fmt.Errorf("state not exist or expired for key=%s", key))
+		}
+		currentEtag := strconv.Itoa(int(item.Kvs[0].ModRevision))
+		if currentEtag != *etag {
+			return state.NewETagError(state.ETagMismatch, fmt.Errorf(
+				"state etag not match for key=%s: current=%s, expect=%s", key, currentEtag, *etag))
+		}
+	}
+	return nil
+}
+
+func (e *Etcd) GetComponentMetadata() map[string]string {
+	metadataStruct := etcdConfig{}
+	metadataInfo := map[string]string{}
+	metadata.GetMetadataInfoFromStructType(reflect.TypeOf(metadataStruct), &metadataInfo)
+	return metadataInfo
+}
+
+func (e *Etcd) Close() error {
+	if e.client == nil {
+		return nil
+	}
+
+	return e.client.Close()
+}
+
+func (e *Etcd) Ping() error {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	if _, err := e.client.Get(ctx, "health"); err != nil {
+		return fmt.Errorf("etcd store: error connecting to etcd at %s: %s", e.client.Endpoints(), err)
+	}
+
+	return nil
+}
+
+// Multi performs a transactional operation. succeeds only if all operations succeed, and fails if one or more operations fail.
+func (e *Etcd) Multi(ctx context.Context, request *state.TransactionalStateRequest) error {
+	if len(request.Operations) == 0 {
+		return nil
+	}
+
+	ops := make([]clientv3.Op, 0)
+	ttls := make([]int, 0)
+	for _, o := range request.Operations {
+		if o.Operation == state.Upsert {
+			req := o.Request.(state.SetRequest)
+			ttlInSeconds, err := e.doSetValidateParameters(&req)
+			if err != nil {
+				return err
+			}
+			ttls = append(ttls, ttlInSeconds)
+		} else if o.Operation == state.Delete {
+			req := o.Request.(state.DeleteRequest)
+			if err := state.CheckRequestOptions(req.Options); err != nil {
+				return err
+			}
+			ttls = append(ttls, 0)
+		}
+	}
+
+	for _, o := range request.Operations {
```

Comment:
Could this loop (and maybe even the next) be performed together with the previous one? Just to loop on the operations one less time